File: ios/JMUDPScoketManager.m
//
//  JMUDPScoketManager.m
//  RNReactNativeFtpJm
//
//  Created by lzj on 2021/1/26.
//  Copyright © 2021 Facebook. All rights reserved.
//

#import "JMUDPScoketManager.h"
#import "JMRealFtpModel.h"
#import <CocoaAsyncSocket/GCDAsyncUdpSocket.h>

NSString *const kRNJMUDPSocketManager = @"listeningUDPScoketCellBack";

@interface JMUDPScoketManager () <GCDAsyncUdpSocketDelegate>

@property (nonatomic,assign) BOOL isHasListeners;

@property (nonatomic,strong) GCDAsyncUdpSocket *socket;
@property (nonatomic,copy) NSString *host;
@property (nonatomic,assign) NSInteger port;
@property (nonatomic,assign) CGFloat sendTimeout;

@end

@implementation JMUDPScoketManager
RCT_EXPORT_MODULE(JMUDPScoketManager);

- (void)startObserving {
    self.isHasListeners = YES;
    [super startObserving];
}

- (void)stopObserving {
    self.isHasListeners = NO;
    [super stopObserving];
}

- (NSArray<NSString *> *)supportedEvents {
    return @[kRNJMUDPSocketManager];    //添加监听方法名
}

- (NSDictionary *)constantsToExport {
    return @{@"kRNJMUDPSocketManager": kRNJMUDPSocketManager};    //导出监听方法名，方便JS调用
}

- (void)sendEventWithCode:(NSInteger)code data:(id)data {
    if (self.isHasListeners) {
        NSMutableDictionary *dic = [NSMutableDictionary dictionary];
        [dic setValue:@(code) forKey:@"code"];
        if (data) {
            [dic setValue:[JMRealFtpModel toJsonString:data] forKey:@"data"];
        }
        
        [self sendEventWithName:kRNJMUDPSocketManager body:[JMRealFtpModel toJsonString:dic]];
    }
}

#pragma mark -

RCT_EXPORT_METHOD(configUDPSocket:(NSString *)host port:(NSInteger)port timeout:(NSInteger)timeout resolver:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter) {
    if (_socket) return;
    self.host = host;
    self.port = port;
    self.sendTimeout = timeout;
    
    _socket = [[GCDAsyncUdpSocket alloc] initWithDelegate:self delegateQueue:dispatch_get_main_queue()];
    NSError *error;
    BOOL ret = [self.socket bindToPort:port error:&error];
    if (!ret || !error) {
        _socket.delegate = nil;
        _socket = nil;
        rejecter(@"811", @"配置参数失败",nil);
        return;
    }
    
    ret = [self.socket enableBroadcast:YES error:nil];
    if (!ret || error) {
        _socket.delegate = nil;
        _socket = nil;
        if (rejecter) rejecter(@"604", @"启动失败",nil);
        return;
    }
    
    [self.socket beginReceiving:&error];
    if (resolver) resolver(nil);
}

RCT_EXPORT_METHOD(send:(NSString *)data tag:(NSInteger)tag resolver:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter) {
    if (!_socket || data) {
        if (rejecter) rejecter(@"605", @"未进行配置参数或data不正确", nil);
        return;
    }
    
    [self.socket sendData:[data dataUsingEncoding:NSUTF8StringEncoding] toHost:self.host port:self.port withTimeout:self.sendTimeout tag:tag];
    if (resolver) resolver(nil);
}

RCT_EXPORT_METHOD(closeSocket:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter) {
    if (_socket) {
        _socket.delegate = nil;
        _socket = nil;
    }
    if (resolver) resolver(nil);
}

#pragma mark - GCDAsyncUdpSocketDelegate

- (void)udpSocket:(GCDAsyncUdpSocket *)sock didNotConnect:(NSError *)error {
    [self sendEventWithCode:601 data:nil];
}

- (void)udpSocketDidClose:(GCDAsyncUdpSocket *)sock withError:(NSError *)error {
    [self sendEventWithCode:602 data:nil];
}

- (void)udpSocket:(GCDAsyncUdpSocket *)sock didConnectToAddress:(NSData *)address {
    [self sendEventWithCode:600 data:nil];
}

- (void)udpSocket:(GCDAsyncUdpSocket *)sock didReceiveData:(NSData *)data fromAddress:(NSData *)address withFilterContext:(id)filterContext {
    __autoreleasing NSError* error = nil;
    id result = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&error];
    if (error != nil)
        return;
 
    id dataDic = [JMRealFtpModel toJsonString:result];
    if (dataDic) {
        [self sendEventWithCode:603 data:nil];
    }
}

@end

File: ios/JMUDPScoketManager.swift
//
// ProjectName:  JMSmallAppEngine
// FileName:     JMUDPScoketManager.swift
// Description:  <#Description#>
//
// Created by LiuLuJia on 2019/11/19.
// COPYRIGHT. ShenZhen JiMi Technology Co., Ltd. 2018.
// ALL RIGHTS RESERVED.
//
// No part of this publication may be reproduced, stored in a retrieval system, or transmitted,
// on any form or by any means, electronic, mechanical, photocopying, recording,
// or otherwise, without the prior written permission of ShenZhen JiMi Network Technology Co., Ltd.
//

import Foundation
import React
import CocoaAsyncSocket
@objc(JMUDPScoketManager)
class JMUDPScoketManager: RCTEventEmitter {
    //发送回调结果给小程序方法名
    let kRNJMUDPSocketManager = "listeningUDPScoketCellBack"
    var isHasListeners = false      //JS是否有监听事件
    var host : String?
    var port : UInt16?
    var sendTimeout : Double?
    var socket : GCDAsyncUdpSocket?
    var sendSuccessCallback : RCTPromiseResolveBlock?
    var sendFailCallback : RCTPromiseRejectBlock?
    var sendDataTag = [Int]()//待发送的指令tag
    deinit {
        debugPrint("UDPScoket->deinit")

    }
    /// JS开启监听事件
    override func startObserving() {
        isHasListeners = true
    }
    
    /// JS停止监听事件
    override func stopObserving() {
        isHasListeners = false
    }
    
    override func supportedEvents() -> [String]! {
        return [kRNJMUDPSocketManager]
    }
    
    override func constantsToExport() -> [AnyHashable : Any]! {
        return ["kRNJMUDPSocketManager": kRNJMUDPSocketManager]
    }
    
    func sendEvent(_ code: Int, _ data: Any? = nil) -> Void {
        if (isHasListeners) {
            var param : [String:Any] = ["code":code]
            if data != nil {
                param["data"] = data!
            }
            let jsonStr = JMFTPSyncFileTools.getJSONStringFrom(param)
            self.sendEvent(withName: kRNJMUDPSocketManager, body: jsonStr)
        }
    }
    //MARK:配置scoket参数
    @objc(configUDPSocket:port:timeout:resolver:rejecter:)
    func configUDPSocket(host:String,
                         port:Int,
                         timeout:Int,
                         resolver:@escaping RCTPromiseResolveBlock,
                         rejecter:@escaping RCTPromiseRejectBlock) {
        self.host = host
        self.port = UInt16(port)
        self.sendTimeout = Double(timeout)
        socket = GCDAsyncUdpSocket(delegate: self , delegateQueue: DispatchQueue.main)
        //监听接口&接收数据
        do {
            try socket?.bind(toPort: UInt16(port))
            try socket?.beginReceiving()
            resolver(nil)

        } catch  {
            rejecter("811","配置参数失败",nil)
        }
        

    }
    //MARK:发送指令
    @objc(send:tag:resolver:rejecter:)
    func send(data:String,
              tag:Int,
              resolver:@escaping RCTPromiseResolveBlock,
              rejecter:@escaping RCTPromiseRejectBlock) {
        guard let socket = self.socket,
            let dataStr = data.data(using: .ascii),
            let host = self.host,
            let port = self.port,
            let sendTimeout = self.sendTimeout else {
                rejecter("605","未进行配置参数或data不正确",nil)
            return
        }
        do {
            try socket.enableBroadcast(true)
            socket.send(dataStr, toHost: host, port: port, withTimeout: sendTimeout, tag: tag)
            sendDataTag.append(tag)
            // receive
            resolver(nil)

        } catch {
            print("dfdsg error=\(error)")
            rejecter("604","发送失败",nil)
        }
    }
    //MARK:关闭socket
    @objc(closeSocket:rejecter:)
    func closeSocket(resolver:@escaping RCTPromiseResolveBlock,
                     rejecter:@escaping RCTPromiseRejectBlock) {
        socket?.close()
        socket = nil
        resolver(nil)
    }
}
extension JMUDPScoketManager: GCDAsyncUdpSocketDelegate{
    //连接失败
    func udpSocket(_ sock: GCDAsyncUdpSocket, didNotConnect error: Error?) {
        debugPrint("didNotConnect tag:\(error.debugDescription)")
        sendEvent(601)
    }
    //发送带有tag数据
    func udpSocket(_ sock: GCDAsyncUdpSocket, didSendDataWithTag tag: Int) {
        debugPrint("didSendDataWithTag tag:\(tag)")
//        if sendDataTag.contains(tag) {
//            sendSuccessCallback?("")
//            sendSuccessCallback = nil
//            sendDataTag = sendDataTag.filter{$0 != tag}
//        }
    }
    //关闭链接
    func udpSocketDidClose(_ sock: GCDAsyncUdpSocket, withError error: Error?) {
        debugPrint("udpSocketDidClose:\(String(describing: sock))")
        sendEvent(602)
    }
    //连接成功
    func udpSocket(_ sock: GCDAsyncUdpSocket, didConnectToAddress address: Data) {
        debugPrint("didConnectToAddress")
        sendEvent(600)
    }
    //发送tag数据失败
    func udpSocket(_ sock: GCDAsyncUdpSocket, didNotSendDataWithTag tag: Int, dueToError error: Error?) {
        debugPrint("didNotSendData error:\(String(describing: error)), tag\(tag)")
//        if sendDataTag.contains(tag) {
//            sendFailCallback?("605","发送数据失败",nil)
//            sendDataTag.remove(at: tag)
//        }
    }
    //接受到数据
    func udpSocket(_ sock: GCDAsyncUdpSocket, didReceive data: Data, fromAddress address: Data, withFilterContext filterContext: Any?) {
        let str = String.init(data: address, encoding: String.Encoding.utf8)
        print("dpSocket(_ sock adress = \(str) filterContext = \(filterContext)")
        if let dict = try? JSONSerialization.jsonObject(with: data, options: .mutableContainers) {
            sendEvent(603,dict)
        }
    }
}
